### PR TITLE
aws_ec2 fix hostnames

### PR DESCRIPTION
**Pull request by alinabuzachis**
##### SUMMARY

aws_ec2- fix hostnames bugs

Should fix https://github.com/ansible-collections/amazon.aws/issues/582 and https://github.com/ansible-collections/amazon.aws/issues/583

1. Fix returned hosts when
```
hostnames:
- tag:Tag1,Tag2
```
2. Fix returned hosts when
```
hostnames:
- tag:Tag1
- tag:Tag2
```
3. Fix returned hosts when
```
hostnames:
- tag:Tag1=Test1,Tag2=Test2
```

Given and EC2 instance with the following tags
```
tags:
  Tag1: Test1
  Tag2: Test2
```
Instead of only returning
```
"aws_ec2": {
    "hosts": [
        "Test1"
    ]
},
"tag_Name_instance_02": {
    "hosts": [
        "Test1"
    ]
},
"tag_Tag1_Test1": {
    "hosts": [
        "Test1"
    ]
},
"tag_Tag2_Test2": {
    "hosts": [
        "Test1"
    ]
}
```
It returns now

```
{
"_meta": {
   "hostvars": {
         "Test1": {
                "ami_launch_index": 0,
                "ansible_host": "10.210.0.101",
                   ...},
         "Test2": {
                 "ami_launch_index": 0,
                 "ansible_host": "10.210.0.101",
                   ...},
    },
    "all": {
        "children": [
            "aws_ec2",
            "tag_Name_instance_02",
            "tag_Tag1_Test1",
            "tag_Tag2_Test2",
            "ungrouped"
        ]
    },
    "aws_ec2": {
        "hosts": [
            "Test1",
            "Test2"
        ]
    },
    "tag_Name_instance_02": {
        "hosts": [
            "Test1",
            "Test2"
        ]
    },
    "tag_Tag1_Test1": {
        "hosts": [
            "Test1",
            "Test2"
        ]
    },
    "tag_Tag2_Test2": {
        "hosts": [
            "Test1",
            "Test2"
        ]
    }
}



```
 

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

aws_ec2

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by alinabuzachis** [comment]
recheck

File: tests/integration/targets/inventory_aws_ec2/templates/inventory_with_hostnames.yml.j2 [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+regions:
+- '{{ aws_region }}'
+keyed_groups:
+- prefix: tag
+  key: tags
+hostnames:
+# can also be specified using
+# - tag:Tag1,Tag2
+# or
+# - tag:Tag1=Test1,Tag2=Test2
```
Comment (tremble) [inline tests/integration/targets/inventory_aws_ec2/templates/inventory_with_hostnames.yml.j2:16]: If possible I'd like to see an integration test of at least one of these.
"tag:Tag1,Tag2" wouldn't be the 'usual' way to write tags in a filter (see #582), so if that leaked into a filter it's more likely to break things than "tag:Tag1,tag:Tag2" appearing in a filter.

Comment (alinabuzachis) [reply tests/integration/targets/inventory_aws_ec2/templates/inventory_with_hostnames.yml.j2]: Added the integration tests for `- tag:Tag1=Test1,Tag2=Test2`.

**Review by tremble — APPROVED** [review]

**Comment by flowerysong** [comment]
This is a massively breaking change, not a bugfix. `hostnames` was designed as an ordered list of hostname candidates, with the first working one used. Switching to using all of them results in the same host being added to the inventory multiple times (#950), which is not what most people want.

I can see why someone might want the new behaviour, but there needs to be a way to turn it off, and it should probably default to off.

With this change:
```
    "az_us_east_2b": {
        "hosts": [
            "ec2-18-221-212-154.us-east-2.compute.amazonaws.com",
            "ec2-3-14-67-171.us-east-2.compute.amazonaws.com",
            "ip-10-0-81-209.us-east-2.compute.internal",
            "ip-10-0-81-33.us-east-2.compute.internal",
            "set-byakhee.jail.x.mail.umich.edu",
            "unimpeded-nunobiki.mx.x.mail.umich.edu"
        ]
    },
```

After reverting to the previous working version:
```
    "az_us_east_2b": {
        "hosts": [
            "set-byakhee.jail.x.mail.umich.edu",
            "unimpeded-nunobiki.mx.x.mail.umich.edu"
        ]
    },
```

**Comment by yamjoepobuda** [comment]
> This is a massively breaking change, not a bugfix.

We had a number of systems break due to this change. We rely on host_vars being a static value. This introduced literal chaos and caused a number of hosts to fallback to role defaults.

**Comment by markuman** [comment]
Maybe a new parameter `unique_hostnames` with `default(true)` must be introduced, that brings the old behaviour back.  
What do you think @alinabuzachis?

**Comment by tremble** [comment]
I'd head in that direction

Might it also be worth exposing "hostname_" group*s* so that all of the hostnames can be accessed?

**Comment by goneri** [comment]
For the record, I'm working on a fix for this, cf https://github.com/ansible-collections/amazon.aws/issues/950

**Pull request merged**